arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/Fold.kt

```diff
@@ -0,0 +1,86 @@
+@file:JvmMultifileClass
+@file:JvmName("Effect")
+@file:OptIn(ExperimentalTypeInference::class)
+
+package arrow.core.continuations
+
+import arrow.core.nonFatalOrThrow
+import kotlin.coroutines.cancellation.CancellationException
+import kotlin.experimental.ExperimentalTypeInference
+import kotlin.jvm.JvmMultifileClass
+import kotlin.jvm.JvmName
+
+/**
+ * `invoke` the [Effect] and [fold] the result:
+ *  - _success_ [transform] result of [A] to a value of [B].
+ *  - _shift_ [recover] from `shifted` value of [R] to a value of [B].
+ *  - _exception_ [error] from [Throwable] by transforming value into [B].
+ *
+ * This method should never be wrapped in `try`/`catch` as it will not throw any unexpected errors,
+ * it will only result in [CancellationException], or fatal exceptions such as `OutOfMemoryError`.
+ */
+public suspend fun <R, A, B> Effect<R, A>.fold(
+  error: suspend (error: Throwable) -> B,
+  recover: suspend (shifted: R) -> B,
+  transform: suspend (value: A) -> B,
+): B = fold({ invoke() }, { error(it) }, { recover(it) }, { transform(it) })
+
+public suspend fun <R, A, B> Effect<R, A>.fold(
+  recover: suspend (shifted: R) -> B,
+  transform: suspend (value: A) -> B,
+): B = fold({ throw it }, recover, transform)
+
+public fun <R, A, B> EagerEffect<R, A>.fold(recover: (R) -> B, transform: (A) -> B): B =
+  fold({ throw it }, recover, transform)
+
+public inline fun <R, A, B> EagerEffect<R, A>.fold(
+  error: (error: Throwable) -> B,
+  recover: (shifted: R) -> B,
+  transform: (value: A) -> B,
+): B = fold({ invoke(this) }, error, recover, transform)
+
+@JvmName("_foldOrThrow")
+public inline fun <R, A, B> fold(
+  @BuilderInference program: Shift<R>.() -> A,
+  recover: (shifted: R) -> B,
+  transform: (value: A) -> B,
+): B = fold(program, { throw it }, recover, transform)
+
+@JvmName("_fold")
+public inline fun <R, A, B> fold(
+  @BuilderInference program: Shift<R>.() -> A,
+  error: (error: Throwable) -> B,
+  recover: (shifted: R) -> B,
+  transform: (value: A) -> B,
+): B {
+  val shift = DefaultShift()
+  return try {
+    transform(program(shift))
+  } catch (e: ShiftCancellationException) {
+    if (e.checkScope(shift)) recover(e.shifted()) else throw e
+  } catch (e: Throwable) {
+    error(e.nonFatalOrThrow())
+  }
+}
+
+/**
+ * **AVOID USING THIS TYPE, it's meant for low-level cancellation code** When in need in low-level code,
+ * you can use this type to differentiate between a foreign [CancellationException] such as JobCancellationException, and the one from [Effect].
+ */
+public class ShiftCancellationException(
```

Comment:
can this be internal or private?

Reply:
This could be `@PublishedApi internal`, meaning that it would still be part of the public API.

The reason I kept it public is that people could use it to distinct `CancellationException` between `JobCancellationException` and `ShiftCancellationException`, but `JobCancellationException` is internal in KotlinX...

TBH I haven't found an actual use-case for this, but it's needs to be public from the Java binary anyway since it appears inside `inline`.

Reply:
I'd make it `@PublishedApi internal` so it's not advertised until we have a use case in Kotlin or need outside the library. By the shape of the type, it seems this is not something people will ever construct manually. The other use case is handling it as an exception which we do ourselves internally.

Reply:
I went a step further, and changed the internals a bit so I could make this type private! 🥳

I agree, since there is not strong use-case to have it public it should not be visible. If such a use-case arises in the future we can ease this restriction.